docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,628 @@
+.. contents::
+    :local:
+    :depth: 2
+
+
+.. testsetup::
+
+    # These are hidden from the docs, but these are necessary for `doctest`
+    # since the `inspect` module doesn't play nicely with the execution
+    # environment for `doctest`
+    import torch
+
+    original_script = torch.jit.script
+    def script_wrapper(obj, *args, **kwargs):
+        obj.__module__ = 'FakeMod'
+        return original_script(obj, *args, **kwargs)
+    torch.jit.script = script_wrapper
+
+    original_trace = torch.jit.trace
+    def trace_wrapper(obj, *args, **kwargs):
+        obj.__module__ = 'FakeMod'
+        return original_trace(obj, *args, **kwargs)
+    torch.jit.trace = trace_wrapper
+
+
+.. _type_system:
+
+
+Type System (I): TorchScript Types
+==================================
+
+This section summarizes TorchScript type system. The information is based on TorchScript language reference
+and reading the implementation.
+
+Terminology
+^^^^^^^^^^^
+
+This document uses the following terminologies:
+
+::
+
+    " " represents real keywords and delimiters that are part of the syntax
+    A | B indicates either A or B
+    ( ) indicates grouping
+    [ ] indicates optional
+    A+ indicates a regular expression where term A is repeated at least once
+    A* indicates a regular expression where term A is repeated zero or more times
+
+TorchScript Types
+^^^^^^^^^^^^^^^^^
+
+The TorchScript type system consists of ``TSType`` and ``TSModuleType`` as defined below.
+
+::
+
+    TSAllType := TSType | TSModuleType
```

Comment:
Meghan's merged PR uses `::=` syntax. Let's use the same syntax for the entire spec and update `:=` to `::=`.